### ISSUE TITLE
Mediation with Unity Ads crashes ( Android only )

### ISSUE THREAD

**Issue by tarouboy** [invertase/react-native-google-mobile-ads#106]
I've tried to add Unity Ads to the mediation but seems there's a problem.

Regarding the official instructions ( [link](https://developers.google.com/admob/android/mediation/unity) ), with Google Mobile Ads SDK ( 20.5.0 ), the tested Unity adapter would be 4.0.0

So I have added below into the build.gradle dependencies:
```
implementation 'com.unity3d.ads:unity-ads:4.0.0'
implementation 'com.google.ads.mediation:unity:4.0.0.0'
```

It's ok to build, however when sending ad requests I found the below Fatal error reported from Flipper:
`com.unity3d.services.UnityServices.initialize() (line:120) :: Error while initializing Unity Services: invalid context, halting Unity Ads init`

Here's what I have tried:
- Update Google Mobile Ads SDK to 20.6.0 with Unity 4.1.0.0 = Same error
- Using older version of Unity SDK 3.6.0.0 which was working with react-native-firebase/ads = same error
- Changing the `delay_app_measurement_init` to either true or false on app.json = Same error

I have been testing with the Meditation Test Suite ( https://github.com/BetOnYou/react-native-mediation-test-suite ), seems SDK and Adapter are both integrated correctly. And it only happens on Android, on iOS it works perfectly.

Anyone else facing similar issue? Thank you!

**Comment by tarouboy** [comment]
Not sure if it helps.

if I set `delay_app_measurement_init` to false
and never run `mobileAds().initialize()`

It works.

**Comment by mikehardy** [comment]
What does a raw `adb logcat` on device show you? Is there some stack trace to work with, with more specific information?

**Comment by tarouboy** [comment]
@mikehardy thanks for your quick reply, here's the log:

```
2022-04-02 00:16:00.398 19776-19975/______ I/UserMessagingPlatform: Use new ConsentDebugSettings.Builder().addTestDeviceHashedId("______") to set this as a debug device.
2022-04-02 00:16:00.406 19776-19975/______ D/UserMessagingPlatform: Stored info not exists: IDFA_freqCapNumViews
2022-04-02 00:16:00.406 19776-19975/______ D/UserMessagingPlatform: Stored info not exists: IABTCF_TCString
2022-04-02 00:16:00.406 19776-19975/______ D/UserMessagingPlatform: Stored info not exists: IABTCF_AddtlConsent
2022-04-02 00:16:00.464 19776-19983/______ E/chromium: [ERROR:simple_version_upgrade.cc(147)] Failed to write a new fake index.
2022-04-02 00:16:00.608 19776-19975/______ D/UserMessagingPlatform: Action[write]: {"IABTCF_CmpSdkID":300,"IABTCF_gdprApplies":0}
2022-04-02 00:16:00.609 19776-19975/______ D/UserMessagingPlatform: Writing to storage: [IABTCF_CmpSdkID] 300
2022-04-02 00:16:00.609 19776-19975/______ D/UserMessagingPlatform: Writing to storage: [IABTCF_gdprApplies] 0
2022-04-02 00:16:00.610 19776-19975/______ D/UserMessagingPlatform: Action[write]: {"IABTCF_idfaFlowControl":2}
2022-04-02 00:16:00.610 19776-19975/______ D/UserMessagingPlatform: Writing to storage: [IABTCF_idfaFlowControl] 2
2022-04-02 00:16:18.659 19776-20014/______ I/DynamiteModule: Considering local module com.google.android.gms.ads.dynamite:0 and remote module com.google.android.gms.ads.dynamite:214106404
2022-04-02 00:16:18.659 19776-20014/______ I/DynamiteModule: Selected remote version of com.google.android.gms.ads.dynamite, version >= 214106404
2022-04-02 00:16:18.660 19776-20016/______ I/DynamiteDebugLogger: [71] com.google.android.gms.ads.dynamite
2022-04-02 00:16:18.675 19776-19872/______ W/______: Loading /data/misc/apexdata/com.android.art/dalvik-cache/arm64/system@framework@com.android.location.provider.jar@classes.odexnon-executable as it requires an image which we failed to load
2022-04-02 00:16:18.704 19776-19872/______ I/DynamiteModule: Considering local module com.google.android.gms.ads.dynamite:0 and remote module com.google.android.gms.ads.dynamite:214106404
2022-04-02 00:16:18.704 19776-19872/______ I/DynamiteModule: Selected remote version of com.google.android.gms.ads.dynamite, version >= 214106404
2022-04-02 00:16:18.705 19776-20016/______ I/DynamiteDebugLogger: [71] com.google.android.gms.ads.dynamite
2022-04-02 00:16:18.707 19776-19872/______ D/DynamitePackage: Instantiated singleton DynamitePackage.
2022-04-02 00:16:18.707 19776-19872/______ D/DynamitePackage: Instantiating com.google.android.gms.ads.ChimeraMobileAdsSettingManagerCreatorImpl
2022-04-02 00:16:18.717 19776-19872/______ I/Ads: Updating ad debug logging enablement.
2022-04-02 00:16:18.724 19776-20021/______ W/Ads: Update ad debug logging enablement as false
2022-04-02 00:16:18.725 19776-20024/______ W/ConnectionStatusConfig: Dynamic lookup for intent failed for action: com.google.android.gms.leibniz.events.service.START
2022-04-02 00:16:18.727 19776-20024/______ W/GmsClient: unable to connect to service: com.google.android.gms.leibniz.events.service.START on com.google.android.gms
2022-04-02 00:16:18.776 19776-19999/______ W/______: Accessing hidden method Landroid/media/AudioManager;->getOutputLatency(I)I (unsupported, reflection, allowed)
2022-04-02 00:16:18.779 19776-19999/______ W/cr_media: Requires BLUETOOTH permission
2022-04-02 00:16:18.795 19776-19776/______ D/UnityAds: com.unity3d.services.core.misc.Utilities.writeFile() (line:127) :: Wrote file: /data/user/0/______/files/UnityAdsStorage-public-data.json
2022-04-02 00:16:18.796 19776-19776/______ D/UnityAds: com.unity3d.services.core.device.Storage.sendEvent() (line:86) :: Couldn't send storage event to WebApp
2022-04-02 00:16:18.796 19776-19776/______ D/UnityAds: com.unity3d.services.ads.UnityAdsImplementation.initialize() (line:72) :: ENTERED METHOD
2022-04-02 00:16:18.796 19776-19776/______ D/UnityAds: com.unity3d.services.UnityServices.initialize() (line:36) :: ENTERED METHOD
2022-04-02 00:16:18.797 19776-19776/______ E/UnityAds: com.unity3d.services.UnityServices.initialize() (line:120) :: Error while initializing Unity Services: invalid context, halting Unity Ads init
2022-04-02 00:16:18.798 19776-19776/______ D/UnityMediationAdapter: {
      "Code": 302,
      "Message": "Unity Ads initialization failed: [INVALID_ARGUMENT] Unity Ads SDK failed to initialize due to invalid context",
      "Domain": "com.unity3d.ads",
      "Cause": "null"
    }
2022-04-02 00:16:18.843 19776-19776/______ I/DynamiteModule: Considering local module com.google.android.gms.ads.dynamite:0 and remote module com.google.android.gms.ads.dynamite:214106404
2022-04-02 00:16:18.843 19776-19776/______ I/DynamiteModule: Selected remote version of com.google.android.gms.ads.dynamite, version >= 214106404
2022-04-02 00:16:18.844 19776-20016/______ I/DynamiteDebugLogger: [71] com.google.android.gms.ads.dynamite
```

**Comment by tarouboy** [comment]
Here's the log when I have turned off `delay_app_measurement`  and commented `mobileAds().initialize()` ( Unity Test Ad shown )

```
2022-04-02 00:26:04.517 20521-20521/______ D/UnityAds: com.unity3d.services.core.misc.Utilities.writeFile() (line:127) :: Wrote file: /data/user/0/______/files/UnityAdsStorage-public-data.json
2022-04-02 00:26:04.519 20521-20521/______ D/UnityAds: com.unity3d.services.core.device.Storage.sendEvent() (line:86) :: Couldn't send storage event to WebApp
2022-04-02 00:26:04.519 20521-20521/______ D/UnityAds: com.unity3d.services.ads.UnityAdsImplementation.initialize() (line:72) :: ENTERED METHOD
2022-04-02 00:26:04.519 20521-20521/______ D/UnityAds: com.unity3d.services.UnityServices.initialize() (line:36) :: ENTERED METHOD
2022-04-02 00:26:04.521 20521-20521/______ I/UnityAds: com.unity3d.services.UnityServices.initialize() (line:131) :: Initializing Unity Services 3.7.5 (3750) with game id ______ in production mode
2022-04-02 00:26:04.527 20521-20521/______ I/UnityAds: com.unity3d.services.UnityServices.initialize() (line:146) :: Unity Services environment check OK
2022-04-02 00:26:04.547 20521-20816/______ I/UnityAds: com.unity3d.services.core.configuration.InitializeThread$InitializeStateConfig.execute() (line:289) :: Unity Ads init: load configuration from https://config.unityads.unity3d.com/webview/3.7.5/release/config.json

```

**Comment by tarouboy** [comment]
@mikehardy after some trial and error, I've changed below code from `getContext()` to `getApplicationContext()` and it seems to solve the problem. 

```
MobileAds.initialize(
        getContext(),
        new OnInitializationCompleteListener() {
          @Override
          public void onInitializationComplete(InitializationStatus initializationStatus) {
            WritableArray result = Arguments.createArray();
            for (Map.Entry<String, AdapterStatus> entry :
                initializationStatus.getAdapterStatusMap().entrySet()) {
              WritableMap info = Arguments.createMap();
              info.putString("name", entry.getKey());
              info.putInt("state", entry.getValue().getInitializationState().ordinal());
              info.putString("description", entry.getValue().getDescription());
              result.pushMap(info);
            }
            promise.resolve(result);
          }
        });
```

**Comment by mikehardy** [comment]
Just a generic sort of "invalid argument" 
https://github.com/googleads/googleads-mobile-android-mediation/blob/9b213e350e8b15513ac97bcb59593d608b6b4bae/ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdsAdapterUtils.java#L131

But it has some info:
https://docs.unity3d.com/Packages/com.unity.ads@4.0/api/UnityEngine.Advertisements.IUnityAdsInitializationListener.html

> UnityAdsInitializationError.INVALID_ARGUMENT, initialization failed due to invalid argument(e.g. game ID)

Have you triple-checked the config used for android? Strange that it does seem to work with different setups (as you mention: delay_app_measurement disabled and no initialize call)

Unfortunately I'm a bit out of my depth here

It looks like the Context is neither an instance of Application or Activity, generating your error message

https://github.com/Unity-Technologies/unity-ads-android/blob/bec9f800f8a16d540f2c883571dce913a5489513/unity-ads/src/main/java/com/unity3d/services/UnityServices.java#L100-L114

That Context argument is coming in here: https://github.com/Unity-Technologies/unity-ads-android/blob/bec9f800f8a16d540f2c883571dce913a5489513/unity-ads/src/main/java/com/unity3d/services/UnityServices.java#L34

Which is from here I think, and what is most strange is that google's unity adapter is checking for a type that is valid: https://github.com/googleads/googleads-mobile-android-mediation/blob/9b213e350e8b15513ac97bcb59593d608b6b4bae/ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java#L134-L137

...assuming you are testing interstial ads, though banner also checks https://github.com/googleads/googleads-mobile-android-mediation/blob/9b213e350e8b15513ac97bcb59593d608b6b4bae/ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityBannerAd.java#L142-L146 as does rewarded https://github.com/googleads/googleads-mobile-android-mediation/blob/9b213e350e8b15513ac97bcb59593d608b6b4bae/ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityRewardedAd.java#L97-L101 and as far as I know you are always getting a ReactActivity object through there which is the correct type.

So...how did it pass through one check and not pass through the following check?

I see your most recent comments (we were working in parallel) and I think the explanation is this:

- the ReactActivity context is actually pretty ephemeral in react-native, it can go null while you are still operating on it in the native code
- the ApplicationContext is long-lived though and more stable, so should work.

If you propose a PR we can check it, and if it works for you in your mediation + the example app still loads ads correctly (https://github.com/invertase/react-native-google-mobile-ads/blob/77a6da374b3753222e49c2ac8c02f95c60d0142c/package.json#L83-L99 - then click around after install/build/packager-start and the app is running) I don't see why we couldn't merge+release

**Comment by tarouboy** [comment]
> If you propose a PR we can check it, and if it works for you in your mediation + the example app still loads ads correctly

I will try then it will be my first PR. lol

**Comment by mikehardy** [comment]
Magic trick: keep this open but in the PR write `Fixes #106` and github will automagically cross-link the PR and the issue, and close this issue when the PR merges, because until then our mediation adapter still has a problem so this should  be open :-)

First PRs merged entitle PR author to all of my internet points for the day :trophy:, yours for the taking! :laughing:

**Comment by mikehardy** [comment]
That's odd - I was not aware that you could close issues in our repo with merges in a fork! That's unexpected, re-opening until the PR is merged here
